google/cloud/bigquery/standard_sql.py

```diff
@@ -43,52 +42,137 @@ class StandardSqlDataType:
-        e.g. INT64, DATE, ARRAY.
+    Args:
+        type_kind:
+            The top level type of this field. Can be any standard SQL data type,
+            e.g. INT64, DATE, ARRAY.
+        array_element_type:
+            The type of the array's elements, if type_kind is ARRAY.
+        struct_type:
+            The fields of this struct, in order, if type_kind is STRUCT.
     """
 
-    array_element_type: Optional["StandardSqlDataType"] = None
-    """The type of the array's elements, if type_kind is ARRAY."""
-
-    struct_type: Optional["StandardSqlStructType"] = None
-    """The fields of this struct, in order, if type_kind is STRUCT."""
-
-    def __post_init__(self):
-        if self.array_element_type is not None and self.struct_type is not None:
+    def __init__(
+        self,
+        type_kind: Optional[
+            StandardSqlTypeNames
+        ] = StandardSqlTypeNames.TYPE_KIND_UNSPECIFIED,
+        array_element_type: Optional["StandardSqlDataType"] = None,
+        struct_type: Optional["StandardSqlStructType"] = None,
+    ):
+        if array_element_type is not None and struct_type is not None:
             raise ValueError(
                 "array_element_type and struct_type are mutally exclusive."
             )
 
-    def to_api_repr(self) -> Dict[str, Any]:
-        """Construct the API resource representation of this SQL data type."""
-
-        if not self.type_kind:
-            type_kind = StandardSqlTypeNames.TYPE_KIND_UNSPECIFIED.value
+        self._properties = {}
+
+        self.type_kind = type_kind
+        self.array_element_type = array_element_type
+        self.struct_type = struct_type
+
+    @property
+    def type_kind(self) -> StandardSqlTypeNames:
+        """The top level type of this field.
+
+        Can be any standard SQL data type, e.g. INT64, DATE, ARRAY.
+        """
+        kind = self._properties["typeKind"]
+        return StandardSqlTypeNames[kind]  # pytype: disable=missing-parameter
+
+    @type_kind.setter
+    def type_kind(self, value: Optional[StandardSqlTypeNames]):
+        new_instance = "typeKind" not in self._properties
+
+        if not new_instance:
+            current = self._properties.get("typeKind")
+
+            if value == current:
+                return  # Nothing to change.
+
+            if (
+                current == StandardSqlTypeNames.ARRAY
+                and self.array_element_type.type_kind
+                is not StandardSqlTypeNames.TYPE_KIND_UNSPECIFIED
+            ):
+                raise ValueError(
+                    "Cannot change {StandardSqlTypeNames.ARRAY} type_kind, first set "
+                    "array_element_type attribute to None."
+                )
+
+            if current == StandardSqlTypeNames.STRUCT and self.struct_type.fields:
+                raise ValueError(
+                    "Cannot change {StandardSqlTypeNames.STRUCT} type_kind, first set "
+                    "struct_type attribute to None."
+                )
```

Comment:
@tswast There's some logic to assure the internal consistency of the `StandardSqlDataType` class, i.e. making sure that `type_kind`, `array_element_type`, and `struct_type` attributes have compatible values.

If you think we should ditch that entirely and treat SQL type classes as merely thin wrappers around the JSON dicts exchanged with the backend, please let me know.

Reply:
I'm in favor of removing client-side validations. There's too much of a risk that we get out-of-sync with the server, otherwise.

Reply:
OK, I'll remove the validation checks and early warnings completely.